Added test for #474

amihaiemil opened the pull request:
Added test ``RealtimeResumeTest.resume_publish_queue()`` to cover #474 

1. connect a ``sender`` and a ``receiver`` (AblyRealtime instances, same as other tests)
2. ``sender`` sends some messages and we check if the ``receiver`` got them
3. We cut the transport of the ``sender`` (disconnect it)
4. We send another batch of messages, which should be queued because ``sender``'s transport is down.
5. Reconnect the ``sender``
6. Check if the receiver got the second batch of messages. 

Other changes: 
- unlike other tests, we use the term ``sender`` and ``receiver`` instead of ``*tx`` and ``*rx`` (makes more sense to me)
- The assertion's messages should be the negation of what we are asserting, because they will apear as errors if the assertion does not pass. Other tests do not use this message correctly...

amihaiemil commented:
@paddybyers After going through the documentation and studying the other tests, I came up with this one. To my understanding, this is the case reported in the Issue. The test seems to pass just fine. 

What do you think?

paddybyers commented on lib/src/test/java/io/ably/lib/test/realtime/RealtimeResumeTest.java:
```diff
@@ -120,7 +120,7 @@ public void resume_simple() {
 			/* wait for the publish callback to be called */
 			ErrorInfo[] errors = msgComplete1.waitFor();
-			assertTrue("Verify success from all message callbacks", errors.length == 0);				
+			assertTrue("Verify success from all message callbacks", errors.length == 0);
```
I assume that your IDE made this change automatically. We do want these changes, but for future reference, please make any such changes (whitespace-only, or style-only) in a separate commit, so we can more easily review the commits that do have code changes.

paddybyers commented on lib/src/test/java/io/ably/lib/test/realtime/RealtimeResumeTest.java:
```diff
@@ -524,4 +524,113 @@ public void resume_verify_publish() {
+			senderChannel.attach();
+			(new ChannelWaiter(senderChannel)).waitFor(ChannelState.attached);
+			assertEquals(
+			    "The sender's channel should be attached!",
```
Just as a minor point of style: we don't need the `!`. Simple, factual, undramatic statements are fine :)

paddybyers commented on lib/src/test/java/io/ably/lib/test/realtime/RealtimeResumeTest.java:
```diff
@@ -524,4 +524,113 @@ public void resume_verify_publish() {
+			e.printStackTrace();
+			fail("init0: Unexpected exception instantiating library");
+		} finally {
+			if(sender != null)
```
As a point of style, I'm not really a fan of a single-line if clause without braces. There are some instances of this in the library and the tests for historical reasons, and we should try to fix them if a file is being modified as part of other work.

paddybyers commented on lib/src/test/java/io/ably/lib/test/realtime/RealtimeResumeTest.java:
```diff
@@ -524,4 +524,113 @@ public void resume_verify_publish() {
+
+			/* wait for the subscription callback to be called */
+			messageWaiter.waitFor(messageCount);
+			assertEquals(
```
In this assertion you're checking that the correct number of messages is received - I think it would have been better to check that the messages received are in fact the same messages as those that were sent. This could be done easily by making the message `name` contain just the index `i`, or `"test_event:" + i` or something like that.

amihaiemil commented:
@paddybyers I've made the following changes

1. Added braces to all the ``if``s in the file which didn't have them.
2. Declared the 2 AblyRealtime instances separately - I find it a smell just like ``if``s without braces :D 
3. Removed the ``!`` from error messages.
4. Enhanced the last assertion as you asked. 

If you don't mind, I'll make a few PRs with cosmetic/style changes when I have some time. For instance, I noticed there are classes which declare their attributes as the last part of the file. It seems to me the original developers were coming from C and learned Java by writing this library :D

paddybyers commented:
> If you don't mind, I'll make a few PRs with cosmetic/style changes when I have some time

In general I'm happy with that, but please don't commit effort to a general update without agreeing the nature of the changes first.

> For instance, I noticed there are classes which declare their attributes as the last part of the file. It seems to me the original developers were coming from C and learned Java by writing this library :D

There was a general approach of keeping private methods and field declarations at the bottom of the class, and public API at the top. That's a legitimate choice, I think, even if it's not your preferred choice. That said, it's never that simple, and sometimes overall organisation and readability is improved if related methods are kept together (including both public and private).

amihaiemil commented:
@paddybyers Thanks for your comments. What happens now? Will this PR be merged? 

About #474, I suggest you close it and open a new one with your thoughts (you said you think the problem is something else). WDYT?
